apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -399,8 +399,13 @@ handle_timeout(_ClientInfo, get_streams, Session) ->
     ensure_timer(get_streams),
     {ok, [], Session};
 handle_timeout(_ClientInfo, bump_last_alive_at, Session0) ->
-    NowMS = now_ms(),
-    Session = session_set_last_alive_at_trans(Session0, NowMS),
+    %% Note: we take a pessimistic approach here and assume that the client will be alive
+    %% until the next bump timeout.  With this, we avoid garbage collecting this session
+    %% too early in case the session/connection/node crashes earlier without having time
+    %% to commit the time.
+    BumpInterval = emqx_config:get([session_persistence, last_alive_update_interval]),
+    EstimatedLastAliveAt = now_ms() + BumpInterval,
+    Session = session_set_last_alive_at_trans(Session0, EstimatedLastAliveAt),
```

Comment:
It could.  I guess it's a question of where this logic should live.  Since this optimization pertains to GC, I think it makes sense to leave it there.
WDYT @keynslug ?

Reply:
To my teste, it's better to store unpreprocessed data in the tables, and do all the calculations on it in the runtime. This way we can change the algorithms more easily without worrying about their side effects.

Reply:
Indeed, if we don't preprocess during this bump, there's no way to differentiate a clean disconnect from the last time it was seen alive (and then crashed).

> Configuration value could change between last update and the start of a GC run.

In this case, it's a bit hard to tell what's the intention of the config update: for the changes to take effect ASAP, or only after the next timestamp bump.

Reply:
I don't think GC should ever make an assumption that `last_alive` is precise. So these problems can be resolved by introducing a large-ish safety margin in the GC.